tests/asynch/basic.rs

```diff
@@ -96,6 +96,43 @@ fn async_spawn_chain() {
     );
 }
 
+#[test]
+fn async_thread_yield() {
+    // This tests if thread::yield_now can be called from within an async block
+    check_dfs(
+        || {
+            asynch::spawn(async move {
+                thread::yield_now();
+            });
+            asynch::spawn(async move {});
+        },
+        None,
+    )
+}
+
+#[test]
+#[should_panic(expected = "assertion failed: `(left == right)`")]
+fn async_atomic() {
+    // This tests if shuttle can correctly schedule a different task
+    // after thread::yield_now is called from within an async block
+    use std::sync::atomic::{AtomicUsize, Ordering};
+    check_dfs(
+        || {
+            let r = Arc::new(AtomicUsize::new(0));
+            let r1 = r.clone();
+            asynch::spawn(async move {
+                r1.store(1, Ordering::SeqCst);
+                thread::yield_now();
+                r1.store(0, Ordering::SeqCst);
+            });
+            asynch::spawn(async move {
+                assert_eq!(r.load(Ordering::SeqCst), 0);
+            });
+        },
+        None,
+    )
+}
+
 #[test]
```

Comment:
I tried doing that, but couldn't get it to act any differently than it would before the PR. Any ideas for this?

Reply:
Something like this fails today but I'd expect it to pass after your PR:

```
#[test]
fn async_mutex() {
    check_dfs(
        move || {
            let lock = Arc::new(Mutex::new(0u64));

            let t1 = {
                let lock = Arc::clone(&lock);
                asynch::spawn(async move {
                    let mut l = lock.lock().unwrap();
                    *l += 1;
                })
            };

            let t2 = asynch::block_on(async move {
                t1.await.unwrap();
                *lock.lock().unwrap()
            });

            assert_eq!(t2, 1);
        },
        None
    )
}
```

Reply:
Oh nice! So this test doesn't actually pass with the current PR, but it passes once I remove `block_after_running` from the execution loop like we discussed. Not totally sure why yet so I'll do a bit more digging.

Reply:
Seems to be that the reason is `lock.lock()` calls `thread::switch()`, then goes into the context switch. In there, `block_after_running` incorrectly blocks because the task is an async and it didn't wake itself up. So overall, not doing `block_after_running` in the context switch is for the best.

Reply:
Yeah, that sounds right. I think moving the block-unless-self-woken check closer to the future logic, rather than baking it into the core of the executor as it is today, is for the best.